### PR TITLE
fix: handling client disconnect gracefully by propagating cancellation exception up to the chat completion handler

### PR DESCRIPTION
**Pull request by adubovik**

### CODE REVIEW

File: tests/test_cancellation.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+            status.set_result(True)
+            raise
+        else:
+            status.set_result(False)
+
+    return _chat_completion
+
+
+@pytest.mark.parametrize("with_heartbeat", [True, False])
+async def test_cancellation(with_heartbeat: bool):
```
Comment (roman-romanov-o) [inline tests/test_cancellation.py]: Please, extend this with the test:
1.Run chat completion
2.Inside of chat completion create nested tasks, that run forever
3.Cancel chat completion
4.Check that nested tasks are also canceled

File: aidial_sdk/utils/_cancel_scope.py [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+    def __init__(self):
+        self._tasks: Set[asyncio.Task] = set()
+        self._on_completed_fut: Optional[asyncio.Future] = None
+        self._cancelling: bool = False
+
+    async def __aenter__(self):
+        return self
+
+    async def __aexit__(self, et, exc, tb):
```
Comment (roman-romanov-o) [inline aidial_sdk/utils/_cancel_scope.py]: Let's use the same naming, that [official PEP](https://peps.python.org/pep-0492/#asynchronous-context-managers-and-async-with) introduces:

```python
    async def __aexit__(self, exc_type, exc, tb):
        ...
```

It's quite easy to confuse `et` with something else, then exception type.

**Review by roman-romanov-o — DISMISSED** [review]

File: aidial_sdk/utils/_cancel_scope.py (hunk 2 above)
Comment (adubovik) [reply aidial_sdk/utils/_cancel_scope.py]: Fixed

File: tests/test_cancellation.py (hunk 1 above)
Comment (adubovik) [reply tests/test_cancellation.py]: Fixed, added more tests

**Review by roman-romanov-o — APPROVED** [review]

**Review by roman-romanov-o — APPROVED** [review]

**Review by vladisavvv — APPROVED** [review]

**Pull request merged**